Add clone functionality for module and containers

benborder opened the pull request:
**Original Issue**: #1110

closes #1110

### Summary
Adds cloning functionality to modules and containers by adding a pure virtual `clone()` method to `Module`.

- `clone()` performs a deep copy of parameters and modules, taking advantage of the underlying tensors copy on write semantics if implemented by the backend in order to minimise memory usage.
- Every module and container now must implement `clone()`. This has been done for the core library and some simpler modules in pkg, the remaining will throw a runtime error indicating clone is unimplmented.
- Core modules have been updated where necessary to add appropriate copy, assignment and move constructors to perform a deep copy of `Variable`.
- Core containers have been updated to use a new macro `FL_BASIC_CONTAINER_CLONING` where possible, which implements `clone()` as well as appropriate copy, assignment and move constructors.
- Users must be aware of and manage the lifetimes and cloning behaviour of their modules/containers. This means if users have any custom or shared lifetime requirements they should not use the `FL_BASIC_CONTAINER_CLONING` macro, but implement their own `clone()` override method as well as copy, assignment and move constructors to achieve their desired behaviour.

### Test Plan (required)
Added some tests and successfully ran locally. Also use CI.

jacobkahn reviewed (commented):
@benborder — this CRTP idea is a nice one. If I'm being honest, I think breaking the existing API is fine, and there's a way we could do this which would be pretty minimally invasive. tl;dr -- I think we can add a pure virtual `clone()` one the base `Module`/`Container` and implement things that way. `Module` already being pure means this is pretty simple as is.

Downstream users that use module but want its copy constructor can and should implement `copy` if they want this functionality. If they don't, they can raise an exception when updating. FL is still in a stage where we're not as concerned about keeping the API perfect -- we're still iterating and have the luxury of making decisions (like this one!). Keeping this as a method on the class avoids some additional inheritance complexity which in my opinion significantly complexifies the interface and type hierarchy.

Let me know what you think. I'll leave more specific code comments after we figure out direction.

benborder commented:
@jacobkahn The intention with this approach was to avoid any breaking changes while minimising the amount of code added/changed. Given breaking changes are an option, the approach you suggest is definitely a better one.

It would be good to get some clarification on the desired behaviour of `Module` and `Container` copy constructors. I'm not sure what use case there is for shallow copy semantics for `Module` and `Container` parameters and modules? So I'd be inclined to require their copy constructors perform a deep copy of parameters/modules, which I think is what you are suggesting?

This would mean simple modules like activation functions can continue to use the compiler generated copy constructor. Furthermore, if the copy constructor is assumed to perform a deep copy there would be no need for a `copy()` member function, only a `clone()`, which would minimise changes to the `Module` interface.

jacobkahn commented:
> desired behaviour of `Module` and `Container` copy constructors

Agree -- let's think about this. My thoughts:
- agree shallow copy via module copy constructors is a bit confusing. Copy-on-write makes sense for tensors and Variables; not modules, since they aren't mutated in the same way. Copy ctor for modules doing a deep copy makes a lot of sense.
- re compiler generated copy ctors -- this can work, but one must be careful, as copying underlying `params_` calls `Variable` copy ctor which shallow copies underlying tensors. That COW is by design and matters because of how autograd works, so I think the implementation of `copy` for modules should be explicit and possibly user defined.

A pure virtual `copy` on `Module` that's implemented by each subclass and used by the copy ctor could make this easy.

jacobkahn commented on flashlight/fl/nn/modules/BatchNorm.cpp:
```diff
@@ -44,6 +44,21 @@ BatchNorm::BatchNorm(
+      featSize_(other.featSize_),
+      numBatchesTracked_(other.numBatchesTracked_),
+      runningMean_(
+          other.runningMean_.tensor().copy(),
```
looks good for now. We can add `Variable::copy` which wraps this down the line.

benborder replied:
Sounds good.

jacobkahn commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -122,6 +129,13 @@ class Module {
+   *
+   * @return A shared pointer of the cloned module.
+   */
+  virtual std::shared_ptr<Module> clone() const = 0;
```
Why have `clone()` return a shared pointer instead of an instance? Returning an instance makes things simpler:

```cpp
Module m;
auto p = std::make_shared<T>(std::move(m.clone()));
```
gives a shared pointer, but it also maintains optionality for the user to do whatever they want with the returned instance, i.e. keep it stack allocated, dump it in a `unique_ptr`, etc.

concretely, that would be:
```cpp
virtual Module clone() const = 0
```

benborder replied:
The standard does not allow an [abstract class](https://en.cppreference.com/w/cpp/language/abstract_class) to be used as a return type unless it is a pointer or reference. Given `shared_ptr` is already used for storing `Modules` in `Containers`, I opted to use `shared_ptr`.

I think there is a strong argument for returning a `unique_ptr` instead, but it would mean changing the implementation of `Container` and any classes which inherit it. Additionally there may be a bunch of flow on effects which need to be considered. However that's beyond the scope of this PR, so it's probably best to stick with `shared_ptr` for now.

jacobkahn reviewed (changes requested):
Looks like CI is failing because some classes are still missing `clone()` implementations -- `flashlight/fl/contrib` modules need this as well. Would be great if you could update `flashlight/pkg` too while you're at it. CI baselines for those coming soon.

benborder commented:
@jacobkahn I went though and fixed up all the compile issues. I added copy constructors and copy assignment operators where the implementation was simple enough or the module/container provided core functionality. For the remaining I added a runtime exception indicating cloning is not implemented.

Something to note. Due to adding copy constructors to perform a deep copy of a modules variables etc, existing code may not function as expected. For example the `ResidualFwd` test in `flashlight/fl/test/contrib/modules/ContribModuleTest.cpp` shares a `BatchNorm` module between `resModule1` and `resModule2` containers. Previously even though a copy was taken when adding the `BatchNorm` module to the containers, the running mean/var variables were updated in forward passes for both containers (as the underlying tensors of the `BatchNorm` module's variables were shared between the 2 networks). With these changes, now if `BatchNorm` is copied, a deep copy is performed for its variables. However, if the previous behaviour is desired, a `shared_ptr` can be used to wrap the module to avoid copies and share the module between multiple modules/containers.

There is one potential issue I foresee with the current implementation of cloning, which is the handling of shared modules between multiple modules/containers. For example, in the `ResidualFwd` test, even though the `BatchNorm` layer is shared between  `resModule1` and `resModule2`, when performing a deep copy of either module their layers will no longer be shared. I think most occurrences are likely to be unintended bugs, but there may be some rare edge case scenarios where it is intended behaviour and as such users will need to handle it accordingly.

jacobkahn commented:
@benborder 

> resnet, prev behavior

Prev behavior would be desired. `shared_ptr` should be used in that case where the copy constructors of modules are used ambiguously.

Any existing user code which uses the weird behavior should be changed anyways, as there's inherent ambiguity in the old system. If you're sharing module state very explicitly, you should use pointers or references, else use the copy ctor.

Tw other things as discussed offline:
- we can and should just return `unique_ptr<Module>` from `clone()`. If the user wants to share the resulting cloned module, they can add `std::shared_ptr<Module>(myModule.clone().release())`. This maintains all of the desired flexibility at negligible overhead cost.
- I'm now coming down on the sides of it not being desirable for tensors to be deep copied in modules. Flashlight has always been firmly supporting copy-on-write with Tensors. The semantics for copying modules, in my opinion, should be as follows:
  - Copying a module deep copies its `Variable` parameters
  - Deep copying a `Variable` simple calls the copy constructor of its underlying `Tensor` rather than pointing to identical tensor data per `sharedData`/`sharedGrad` and friends.
  - Underlying tensor library implementations can track whether or not tensors need to be deep copied based on the operations that will be performed on them. For example: of a module is being copied but its parameters only being read from, the tensors underlying its variables don't need to be deep copied, and can be read from the same data.

benborder commented:
After offline discussion with @jacobkahn about if Modules should fully deep copy their parameters circumventing COW semantics, or deep copy using COW semantics if implemented in the backend, identifying some potential edge case issues and resolving some misunderstandings on my part, we agreed:
- Whilst cloned modules/containers that are using tensor backends with COW semantics are not thread safe, this can either be addressed by the user by adding synchronisation, or resolved in the future by providing functionality to allow a full deep copy or even by some other means.
- It will be up to users to correctly manage the lifetimes and cloning behaviour of their modules/containers and be aware of any consequences of cloning a module/container. This means if users have any custom or shared lifetime requirements they should not use the `FL_BASIC_CONTAINER_CLONING` macro, but implement their own `clone()` override method as well as copy, assignment and move constructors to achieve their desired behaviour.

The following changes were made:

- Return `unique_ptr` for `clone()` instead of `shared_ptr`. `unique_ptr` can be promoted to a `shared_ptr` and a convenience function was added to `Container` to facilitate this.
- Added a copy method to `Variable` with optional forced deep copy to circumvent COW of the underlying tensor if applicable.
- Added missing move constructors.
- Ensure train is correctly set in `Module`.

jacobkahn reviewed (changes requested):
Leaving some preliminary comments first — will in parallel start reviewing everything else.

benborder commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -52,8 +52,11 @@ class Module {
    *
    * @param params a vector of `Variable` which will replace `params_`
+   * @param train Sets the module to training mode if true, eval mode if false.
+   * This changes all parameters so that gradient calculation will be
+   * enabled/disabled for any calls to `forward`.
    */
-  explicit Module(const std::vector<Variable>& params);
+  explicit Module(const std::vector<Variable>& params, bool train = true);
```
This was necessary to correctly handle custom copy constructors of derived modules. If a `Module` has train set to false and is then copied, the copied module would have train set to true, but the copied parameters would have their `sharedGrad_->calcGrad` set to false.

I did consider just setting the params to the default train = true in the constructor, but this would break cloning containers where some modules have gradients disabled. For example a pre-trained feature extractor module/container as part of a larger network.

jacobkahn replied:
I feel the better way to do this would be to call `module.eval()` on the newly-constructed module if the original instance that is being copied is in train mode. Should be simple:
```cpp
if (other.isTrain()) {
  inst.train();
}
// symmetrically for eval
```

I'm also now realizing there's no way to _read_ whether a module is in train mode or not. Would you be willing to send a separate PR adding that and a few small tests?

benborder replied:
My concern is that if a user does not call `train()` or `eval()` after clone/copy for whatever reason, it allows the module to be in an invalid state such that the modules parameters `sharedGrad_->calcGrad` and `train_` can be inconsistent after cloning. For example:

```cpp
Linear lin(10,20);
lin.eval();

auto linCopy = lin; // linCopy params gradients are disabled, but train_ = true
// now either train() or eval() must be called depending on the desired behaviour
```
Hence *every* time a copy is made (via clone or via C++ value semantics like above) the user must ensure the correct train mode is set. If it is not done, it could lead to subtle bugs where gradients aren't correctly propagated and its hard to identify the cause, or gradients are collected at inference unnecessarily.

I think it also just makes sense that a cloned/copied module should be in an identical state as its original and shouldn't require a user to take action to ensure it is. I think it's also a good idea to prevent objects from entering an invalid state under normal use.

benborder replied:
Perhaps an alternative approach might resolve this? It would essentially involve checking the parameters and setting `train_ = true` if any parameters have gradients enabled, so something like this:

```cpp
Module::Module(const std::vector<Variable>& params)
    : params_(params.begin(), params.end()),
        train_(std::any_of(params.begin(), params.end(),
            [](const Variable& param) { return param.isCalcGrad(); })
        ) {}
```

This has the advantage of ensuring the module is not in an invalid state when constructed, avoids adding `train` as a constructor parameter and does not call `eval()` or `train()` in the constructor. Users can then call `eval()` or `train()` based on their use case.

Thoughts on this approach?

benborder replied:
Following on from the other comment chain regarding `cloneModule`, I think maybe we've gotten lost in the weeds. Given that the constructor `Module::Module(const std::vector<Variable>& params)` is protected and only derived modules can call it, the derived module can also just set `train_` accordingly in its constructor. For example see `fl/nn/modules/Linear.cpp`:

```cpp
Linear::Linear(const Linear& other)
    : UnaryModule(other.copyParams()),
      nIn_(other.nIn_),
      nOut_(other.nOut_),
      bias_(other.bias_) {
  train_ = other.train_; // <----- train is set here
}
Linear& Linear::operator=(const Linear& other) {
  params_ = other.copyParams();
  train_ = other.train_; // <----- train is set here
  nIn_ = other.nIn_;
  nOut_ = other.nOut_;
  bias_ = other.bias_;
  return *this;
}
```

This can also be done in the copy assignment operator for the derived Module. I've originally updated most Modules in flashlight to do this in the PR. While reviewing the PR changes, I found a few places where I missed it, hence my intention with setting `train_` in the constructor of `Module` was to try remove this source of bug where users forget to set it in their derived Module copy constructor. However, I've just realised this doesn't address the copy assignment operator case. So given users need to copy `params_` in their copy constructors anyway, I believe it would be better to be more consistent and simply require users also set `train_` in both their copy constructors and copy assignment operators and leave the `Module` constructor as it was originally.

benborder commented on flashlight/fl/autograd/Variable.h:
```diff
@@ -114,6 +114,15 @@ class Variable {
+   * tensor's copy constructor behaviour, which is backend dependent.
+   * @return returns the cloned and detached variable.
+   */
+  Variable copy(bool forceDeepCopy = false) const;
```
That makes sense. The primary reasoning for this was for convenience as it would reduce the amount of boilerplate conditional statements in modules that may want to explicitly perform a deep copy. Given users can still achieve deep copy behaviour via `Variable(x.tensor().copy(),x.isCalcGrad())`, its not technically adding any new functionality, just convenience.

benborder commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -122,6 +141,13 @@ class Module {
+   *
+   * @return A unique pointer of the cloned module.
+   */
+  virtual std::unique_ptr<Module> clone() const = 0;
```
This would only be required if the module is a base pointer, otherwise value copy semantics can be used. Where a `cloneModule()` function would be useful is when cloning via calling the base `Module` virtual clone method from a smart pointer. So I could add something like this:

```cpp
template <typename T, typename D, template <typename, typename> class SmartPtr>
SmartPtr<T, D> cloneModule(const SmartPtr<Module, D>& module) {
  auto copy = module->clone();
  return SmartPtr<T, D>(dynamic_cast<T*>(copy.release()), std::move(copy.get_deleter()));
}
```
Which would be used like:
```cpp
std::unique_ptr<Module> modulePtr = std::make_unique<Linear>(10, 20);
std::unique_ptr<Linear> clonedPtr = cloneModule<Linear>(modulePtr);
// or
std::shared_ptr<Module> modulePtr = std::make_shared<Linear>(10, 20);
std::shared_ptr<Linear> clonedPtr = cloneModule<Linear>(modulePtr);
```

jacobkahn replied:
This implementation with smart pointers makes sense.

As a corollary to this, though, (and playing devil's advocate for a moment) what if we considered using a `cloneModule` function that had overloads for most container types (in addition to derived types of `Module`), and removed the `clone()` function on the `Module` interface entirely, using that instead?

benborder replied:
Removing `clone()` from `Module` and replacing with a free standing `cloneModule` function with overrides for all module implementations will not solve the type erasure issue when a pointer to the base `Module` is used (for example in `Container` or in `WeightNorm`). The derived type is unknown at compile time, hence the correct copy constructor can't be called:

```cpp
std::unique_ptr<Module> linearPtr = std::make_unique<Linear>(10, 20);
Linear linearCopy = *linearPtr; // **compile time error** unless the module is cast to its derived type, but we may not know the type at compile time
```
If the type is known at compile time:
```cpp
std::unique_ptr<Linear> linearPtr = std::make_unique<Linear>(10, 20);
Linear linearCopy = *linearPtr; // Copy via copy ctor as the derived type is known at compile time
std::unique_ptr<Linear> clonedPtr = cloneModule<Linear>(linearPtr); // or via the proposed cloneModule function
```

benborder replied:
I think there may be some confusion here (or perhaps I'm misunderstanding?). The `cloneModule<T>()` function does not provide a compile time-guaranteed-safe method of cloning, its purely a convenience function. The type is erased at compile time and run time, hence it is possible for the user to specify the wrong type when casting and because the actual type is only known at run time, at the very least a dynamic cast must be used. Otherwise using the wrong type leads to UB.

> An alternative idea might be to make Module::clone protected and cloneModule a friend, though the templates are tricky.

I dont believe this is possible as the template parameter for `cloneModule<T>` would be unknown to `Module`.

> Orthogonally -- could we not have a cloneModule implementation that operates on stack allocated modules as well or takes a const T& where we static assert std::is_base_of<Module, T>::value?

If a module is stack allocated its type must be known, hence the copy constructor can be used instead of using `Module:clone()`. `Module:clone()` is only necessary when the type has been erased or if it is desired to obtain a type erased clone.

I suspect the real problem here is that there is no reflection in C++ (both compile and run time). So common functionality used in python (e.g. metaclasses) is not possible at this stage. Compile time reflection may be added in C++26, but at the rate its going its unlikely to land.

jacobkahn replied:
Good points here, and of course safety isn't guaranteed given that there's a cast, but that's a constraint of the return type of `clone()` being `Module`.

I see there still being merits for `cloneModule` -- low hanging fruit it saves the user a cast from `Module` (returned by `clone`) back to the base type. It's the difference between:
```cpp
std::unique_ptr<Linear> linear = ...;
std::unique_ptr<Linear> linear2 = cloneModule(linear);
// versus
std::unique_ptr<Linear> linear2 = std::dynamic_pointer_cast<Linear>(linear->clone());
```
with the benefit that the correctness of the cast is guaranteed (though the cast is still required, of course), meaning `static_pointer_cast` can be used, with assurances. This saves a bit of overhead/removes some ugliness.

> dont believe this is possible as the template parameter for cloneModule<T> would be unknown to Module

Template friends mean that any that any T for ` cloneModule<T>` that's a friend of `Module` would be accepted, so this approach is viable ([if my understanding is correct](https://en.cppreference.com/w/cpp/language/friend#Template-friends)). If we made `clone` protected, in this case, there would be no dynamic casts present in end-to-end usage of the API, which in my view would be much less error prone.

benborder replied:
> I see there still being merits for cloneModule -- low hanging fruit it saves the user a cast from Module

Yep, completely agree with this. My perspective is that its purely a convenience function.

> Template friends mean that any that any T for cloneModule<T> that's a friend of Module would be accepted, so this approach is viable

Ah I misunderstood what you were trying to achieve regarding templated friend functions, you are indeed correct that its viable! Though, perhaps I'm still misunderstanding what you are trying to achieve, as I'm not sure what making `cloneModule` a friend would do that using the public interface of `Module` wouldn't? Unless you are thinking of setting `train_` for the cloned/copied module?

> If we made clone protected, in this case, there would be no dynamic casts present in end-to-end usage of the API, which in my view would be much less error prone.

I'm not following this statement as based on my understanding if the derived type is known at compile time, then the copy constructor can be called rather than calling `cloneModule()` or `Module::clone()`. If the derived type is unknown at compile time, then we must necessarily use dynamic_cast to ensure the module is being cast to the correct type.

Thinking about this more and in regards to an earlier comment about using `cloneModule` as a single source for cloning, I don't believe it's possible as the copy constructor for derived modules can also be called. I'm not sure if there is a reasonable approach to enforce making the copy constructor of derived modules protected without completely changing the design of this PR and a lot of Flashlight itself (i.e. using the macro approach that PyTorch C++ API implements).

jacobkahn replied:
After playing with this myself finally, I agree with your perspective. To summarize:
1. We need `clone()` no matter what because we can't have virtual ctors (`clone` is literally the idiom, of course)
2. It's necessary that we express that an abstract `Module` instance _is_ copyable if the derived type isn't known and provide a path to doing so
3. `cloneModule` is only useful insofar as it works on abstract `Module`s. We don't know the derived type, so we have a dynamic cast no matter what.

Overall, since `cloneModule` really just hides `dynamic_cast`, I'd actually rather we remove it (I was wrong!) and have the user do dicey things with casting instead of condoning it in an API. I'm sorry to make you spin your wheels on this; thanks for sticking with me on it :)

benborder replied:
Under the right circumstance `cloneModule` could be valuable to have, but I do agree that it hides dynamic_cast and does give users more tools to shoot themselves in the foot. If a user does find it useful to have, its such a small function, they can easily implement it themselves.

benborder commented on flashlight/pkg/speech/criterion/Seq2SeqCriterion.cpp:
```diff
@@ -127,6 +127,11 @@ Seq2SeqCriterion::Seq2SeqCriterion(
   setUseSequentialDecoder();
 }
 
+std::unique_ptr<Module> Seq2SeqCriterion::clone() const {
```
I'm not familiar enough with these use cases or code. It appears to have a more complex shared module lifetime, due to the `attentions` vector of `shared_ptr` modules provided in its constructor. It also has a `std::shared_ptr<WindowBase> window_;` member, which may make sense to shallow copy?

So although sole ownership is used for `attention` based on `src/flashlight/app/asr/Train.cpp`, I'm not familiar enough with the code or use case to make a call one way or another, so I left it unimplemented for now. It was the same with a few others as well.

If you want I can just assume sole ownership and deep copy everything?

jacobkahn replied:
Nah, making this non-copyable is fine for now. If someone elsewhere is copying it, they can add a copy ctor. We won't get to everything now. We can fix this later at worst.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,39 @@ class Container : public Module {
    * @param module the module to add.
    */
   template <typename T>
-  void add(const T& module) {
```
Yeah, I think making copy an explicit operation is generally a better approach. Using an rvalue reference also enables objects to be constructed or cloned outside and then moved into the container. I think overall this is a much more explicit and flexible approach.

jacobkahn commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -96,6 +106,15 @@ class Module {
+   * `Variable::copy` for more details about behaviour.
+   * @return a copy of the modules parameters as a vector of `Variable`
+   */
+  virtual std::vector<Variable> copyParams(bool forceDeepCopy = false) const;
```
Might've left this in an earlier comment (apologies if I did; I can't find it) -- I think `forceDeepCopy` will be used infrequently enough to justify not including it as an option here. If users actually want to force an _eager_ deep copy, that's specialized behavior that they should implement themselves.

benborder replied:
Yep, above, its related to the `Variable::copy()` comment, I agreed to remove it.

jacobkahn reviewed (changes requested):
It might make sense to add a small test fixture to test copying basic modules (make sure behavior is the same; this can also serve as a chance to test that parameters aren't shared) as well as a test for copying a basic compositional container. Would you be able to add those in this PR as well?

benborder commented on flashlight/fl/examples/RnnLm.cpp:
```diff
@@ -126,6 +126,10 @@ class RnnLm : public Container {
     return "RnnLm";
   }
 
+  std::unique_ptr<Module> clone() const {
```
Fixed.

benborder commented on flashlight/fl/nn/modules/WeightNorm.cpp:
```diff
@@ -49,6 +64,21 @@ void WeightNorm::computeWeight() {
 }
 
+void WeightNorm::initParams() {
+  auto module_params = module_->params();
```
The `transformData()` method is called prior to this, which performs checks. I've add a bounds check to the first access, just to make sure.

benborder commented on flashlight/fl/nn/modules/WeightNorm.cpp:
```diff
@@ -49,6 +64,21 @@ void WeightNorm::computeWeight() {
 }
 
+void WeightNorm::initParams() {
+  auto module_params = module_->params();
```
This code was moved from the header to the implementation file, I was keeping it unmodified to minimise changes. I've now changed it to match the convention.

benborder commented:
I had copy/cloning tests incorporated into existing module tests. I think it makes sense to separate them out to make it clearer/explicit.

jacobkahn reviewed (commented):
This looks great, and I really appreciate the continued patience (with my slowness) and deliberate discussion. Adding a few comments and follow-ups.

benborder commented on flashlight/fl/autograd/Variable.cpp:
```diff
@@ -84,6 +84,10 @@ Tensor& Variable::tensor() const {
 }
 
+Variable Variable::copy() const {
+  return Variable(sharedData_->data, sharedGrad_->calcGrad);
```
No worries 👍

benborder commented on flashlight/fl/contrib/modules/PositionEmbedding.h:
```diff
@@ -23,10 +23,18 @@ namespace fl {
  *
  */
-class PositionEmbedding : public Container {
+class PositionEmbedding : public Module {
```
It was unnecessary to inherit from `Container` as `PositionEmbedding` does not compose any other modules or use any inherited functionality from `Container`. It simply transforms the inputs in the forward pass. It should probably be a `UnaryModule`, but that will change its `forward()` function signature.

benborder commented on flashlight/fl/contrib/modules/Residual.h:
```diff
@@ -135,6 +135,8 @@ class Residual : public Container {
 
   std::string prettyString() const override;
+
+  FL_BASIC_CONTAINER_CLONING(Residual)
```
Yep, its purely for convenience. If not included, the copy constructor, copy assignment operator, move constructor, move assignment operator and clone method must be implemented manually.

benborder commented on flashlight/fl/contrib/modules/TDSBlock.cpp:
```diff
@@ -53,13 +53,13 @@ TDSBlock::TDSBlock(
   }
 
-  add(conv);
+  add(std::move(conv));
```
Yeah, it should become a lot more common now with the changes in this PR.

benborder commented on flashlight/fl/examples/RnnClassification.cpp:
```diff
@@ -147,21 +147,49 @@ class RnnClassifier : public Container {
-                RnnMode::GRU,
-                0 /* Dropout */)),
-        linear_(Linear(hiddenSize, numClasses)),
+      : embed_(std::make_shared<Embedding>(hiddenSize, vocabSize)),
```
Yeah, making object lifetimes more explicit really helps.

benborder commented on flashlight/fl/examples/RnnLm.cpp:
```diff
@@ -79,41 +79,65 @@ class LMDataset : public Dataset {
+    createLayers();
+  }
+
+  RnnLm(RnnLm&& other) = default;
```
Correct, the move constructor and assignment operator aren't required. In the example the only one that is required is the copy constructor as its used by `clone()`. This example was not updated to correctly deep copy its members when cloned in order to preserve the previous behaviour. Given its an example, perhaps it should be updated to demonstrate the correct way to handle writing a `Container`?

jacobkahn replied:
It might make sense to rely on tests or a standalone example to show that instead, don't worry about some of these examples. They're rather old.

benborder commented:
@jacobkahn It seems one of the CircleCI runners bugged out, are you able rerun it?

jacobkahn commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -177,6 +200,38 @@ class BinaryModule : public Module {
+ * @return The new cloned T module pointer
+ */
+template <typename T, template <typename, typename> class SmartPtr, typename D>
+std::unique_ptr<T> cloneModulePtr(const SmartPtr<Module, D>& modulePtr) {
```
To summarize above discussion — let's just remove `cloneModulePtr` and `cloneModuleAs`. We can test `Module::clone()` is isolation, cast in tests, and go from there

benborder replied:
Sounds good. It will be easy to add later if desired.

jacobkahn reviewed (changes requested):
Final stretch! Just a few more comments. Thanks for bearing with me on this. I'll rebase this branch shortly.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,45 @@ class Container : public Module {
+
+  /**
+   * Find orphaned params (i.e. params not in modules contained in the modules_
+   * list).
```
Added a quick example.

Honestly, the main reason for multimap is it has less boilerplate. Using `std::unordered_map<int, std::vector<int>>` would require to first call `find(index)` then check the iterator is not null, then dereference and iterate over the vector. Whereas multimap::equal_range just returns 2 iterators (begin and end) and you can iterate over the range. Its also a lot easier to use when adding elements to the multimap compared to the unordered_map of vectors.

Also I wasn't too concerned with performance here as there are unlikely to be more than 10 orphaned params (let alone 100), cloning a container is unlikely to be a hot loop and even if it was this is unlikely to be any bottleneck.

I do understand that `std::unordered_map<int, std::vector<int>>` might be a more familiar interface, so I'm happy to switch to using it if its preferred.

benborder commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -63,6 +65,13 @@ class Module {
   std::vector<Variable> params() const;
 
+  /**
+   * Gets the nunber of parameters of the module.
```
I see how confusion might arise, change to `numParamTensors`, though would `numParamVariables` be more correct?

benborder commented on flashlight/pkg/vision/nn/Transformer.cpp:
```diff
@@ -416,10 +509,35 @@ Transformer::Transformer(
+Transformer::Transformer(const Transformer& other) {
+  copy(other);
+  createLayers();
+}
```
I ran check format and it didn't make any changes, so I inserted a newline manually.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,45 @@ class Container : public Module {
   template <typename T>
-  void add(const T& module) {
-    add(std::make_shared<T>(module));
+  void add(std::unique_ptr<T>&& module) {
```
Ah yep, good catch. Changed.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,45 @@ class Container : public Module {
-  void add(const T& module) {
-    add(std::make_shared<T>(module));
+  void add(std::unique_ptr<T>&& module) {
+    static_assert(
```
Removed.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,45 @@ class Container : public Module {
+   */
+  template <typename T>
+  void add(T&& module) {
+    static_assert(
```
In this case the static_assert is required as the T could be an lvalue reference and not an rvalue reference. The module would be moved but its not clear to the user that it was. The static_assert ensures that a move is explicitly used. For example:
```cpp
Linear ln(10,20);
my_container.add(ln); // Without the static_assert this compiles without error
ln.getParam(0); // <-- The move is not destructive, but now ln is empty, but there is no way to know
```
Hence users must either use an explicit move or an rvalue:
```cpp
Linear ln(10,20);
my_container.add(std::move(ln)); // move is explicit
my_container.add(Linear(10,20)); // rvalue
```

benborder commented on flashlight/fl/nn/modules/Module.h:
```diff
@@ -177,6 +200,38 @@ class BinaryModule : public Module {
+ * Casts the modulePtr to T* and calls the copy constructor to clone the module,
+ * returning a smart pointer.
+ *
+ * @param modulePtr The module pointer to clone
```
Sure, but I've not seen it used anywhere else in the codebase.

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -117,6 +146,34 @@ class Container : public Module {
+ * Adds a copy constructor, copy assignment operator, move constructor, move
+ * assignment operator and clone method to the class. This should only be used
+ * if the class basically acts as a wrapper around a container such that no
+ * custom module ownership is used. Users should implement these methods
```
Done.

benborder commented on flashlight/fl/test/nn/ModuleTest.cpp:
```diff
@@ -22,6 +22,34 @@ namespace {
+  ContainerTestClass& operator=(ContainerTestClass&& other) = default;
+  void copy(const ContainerTestClass& other) {
+    auto orphanParamIdxMap = other.getOrphanedParamsIdxMap();
+    for (int i = -1; i < static_cast<int>(other.modules_.size()); ++i) {
```
`i` needs to be a negative for this loops to function as intended, hence I believe it needs to be `int`. In the future if updating to C++20, this could be changed to use `ssize()`.

jacobkahn reviewed (approved):
🎉 I think we're ready to merge this after
- rebasing
- CI still being green enough
- double checking any new non-class functions are explicitly exported via FL_API (I don't think there are any)

Cheers to sticking with this OR -- this makes so many semantics clearer and behavior more explicit. A huge step towards a less bug prone set of interfaces on modules, and copy semantics throughout FL!

benborder commented on flashlight/fl/nn/modules/Module.cpp:
```diff
@@ -57,6 +66,10 @@ std::vector<Variable> Module::params() const {
   return params_;
 }
 
+int Module::numParamTensors() const {
```
An int type was used in `Modules` and `Containers` as the index arguments for params, so I followed suit. Using the same type may avoid having to cast in some scenarios.

benborder commented on flashlight/fl/nn/modules/Module.cpp:
```diff
@@ -33,6 +33,15 @@ void Module::setParams(const Variable& var, int position) {
 }
 
+std::vector<Variable> Module::copyParams() const {
+  std::vector<Variable> params;
```
`std::transform` will only replace the for loop, which can also easily be interpreted as a one liner:
```cpp
for (const auto& param : params_) { params.emplace_back(param.copy()); }
```
For the `std::transform` version, when formatted, it actually ends up being 4 lines vs the 3 with the for loop. In my opinion I think it's also harder to understand the std::transform version at a glance.
```cpp
std::transform(
    params_.begin(), params_.end(), params.begin(), [](const auto& param) {
      return param.copy();
    });
```

benborder commented:
Awesome!

Agreed, it's a solid set of improvements. Glad we can get it over the line!

jacobkahn commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -117,6 +167,47 @@ class FL_API Container : public Module {
+  ContainerClass& operator=(ContainerClass&& other) = default; \
+  std::unique_ptr<Module> clone() const override {             \
+    return std::make_unique<ContainerClass>(*this);            \
+  }
```
this macro doesn't need a semicolon when used, fwiw

benborder replied:
Ah yep, fixed.

jacobkahn commented on flashlight/fl/examples/RnnClassification.cpp:
```diff
@@ -151,21 +151,49 @@ class RnnClassifier : public Container {
+  RnnClassifier& operator=(const RnnClassifier& other) {
+    clear();
+    copy(other);
+    createLayers();
```
copy assignment operator missing a return value

benborder replied:
Oops. I guess the examples aren't compiled in the CI now (if they ever were)?

jacobkahn commented on flashlight/fl/examples/RnnClassification.cpp:
```diff
@@ -151,21 +151,49 @@ class RnnClassifier : public Container {
+  }
+
+  std::unique_ptr<Module> clone() const override {
+    throw std::make_unique<RnnClassifier>(*this);
```
I think this is supposed to be `return`? :)
```suggestion
    return std::make_unique<RnnClassifier>(*this);
```

benborder replied:
Yeah, find replace got me!

jacobkahn commented on flashlight/fl/examples/RnnClassification.cpp:
```diff
@@ -103,10 +103,10 @@ class ClassificationDataset : public Dataset {
     for (auto& l : lang) {
       read(datasetPath, l);
-}
+    }
```
would you mind adding `explicit` to the `ClassificationDataset` constructor too? :D

benborder replied:
Sure!

jacobkahn commented on flashlight/fl/examples/RnnLm.cpp:
```diff
@@ -79,41 +79,66 @@ class LMDataset : public Dataset {
+    createLayers();
+  }
+
+  RnnLm(RnnLm&& other) = default;
```
the default the compiler uses should be fine, no? This line shouldn't be needed?

benborder commented on flashlight/fl/nn/modules/Container.h:
```diff
@@ -40,16 +40,66 @@ class FL_API Container : public Module {
+    static_assert(
+        !std::is_lvalue_reference_v<T>,
+        "add() can only accept rvalues. Use std::move().");
+    add(std::make_shared<std::decay_t<T>>(std::move(module)));
```
I believe as the static assert checks for an lvalue reference, `std::forward<T>` isn't strictly required and `std::move` is sufficient - or rather in this case they function the same. I think I used `std::move` instead as it was slightly less to type. I'll change to avoid confusion.
